lightning/src/ln/channelmanager.rs

```diff
@@ -3325,15 +3406,46 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		} else { unreachable!(); }
 	}
 
+	fn finalize_claims(&self, mut sources: Vec<HTLCSource>) {
+		for source in sources.drain(..) {
+			if let HTLCSource::OutboundRoute { session_priv, payment_id, .. } = source {
+				let mut session_priv_bytes = [0; 32];
+				session_priv_bytes.copy_from_slice(&session_priv[..]);
+				let mut outbounds = self.pending_outbound_payments.lock().unwrap();
+				if let hash_map::Entry::Occupied(mut payment) = outbounds.entry(payment_id) {
+					assert!(payment.get().is_fulfilled());
+					payment.get_mut().remove(&session_priv_bytes, 0); // Note that the amount is no longer tracked
+					if payment.get().remaining_parts() == 0 {
+						payment.remove();
+					}
+				}
+			}
+		}
+	}
+
 	fn claim_funds_internal(&self, mut channel_state_lock: MutexGuard<ChannelHolder<Signer>>, source: HTLCSource, payment_preimage: PaymentPreimage, forwarded_htlc_value_msat: Option<u64>, from_onchain: bool) {
 		match source {
 			HTLCSource::OutboundRoute { session_priv, payment_id, path, .. } => {
 				mem::drop(channel_state_lock);
 				let mut session_priv_bytes = [0; 32];
 				session_priv_bytes.copy_from_slice(&session_priv[..]);
 				let mut outbounds = self.pending_outbound_payments.lock().unwrap();
-				let found_payment = if let Some(mut sessions) = outbounds.remove(&payment_id) {
-					sessions.remove(&session_priv_bytes, path.last().unwrap().fee_msat)
+				let found_payment = if let hash_map::Entry::Occupied(mut payment) = outbounds.entry(payment_id) {
+					let found_payment = !payment.get().is_fulfilled();
+					payment.get_mut().mark_fulfilled();
+					if from_onchain {
+						// We currently immediately remove HTLCs which were fulfilled on-chain.
+						// This could potentially lead to removing a pending payment too early,
+						// with a reorg of one block causing us to re-add the fulfilled payment on
+						// restart.
+						// TODO: We should have a second monitor event that informs us of payments
+						// irrevocably fulfilled.
+						payment.get_mut().remove(&session_priv_bytes, path.last().unwrap().fee_msat);
+						if payment.get().remaining_parts() == 0 {
+							payment.remove();
+						}
```

Comment:
I'm pretty confident in this code path, but FWIW all tests pass with these commented out